internal/target/repository.go

```diff
@@ -504,18 +475,30 @@ func (r *Repository) UpdateTarget(ctx context.Context, target Target, version ui
 	if target == nil {
 		return nil, nil, nil, db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "missing target")
 	}
+	if target.GetPublicId() == "" {
+		return nil, nil, nil, db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "missing target public id")
+	}
+	if version == 0 {
+		return nil, nil, nil, db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "missing version")
+	}
 	vet, ok := subtypeRegistry.vetForUpdateFunc(target.GetType())
 	if !ok {
 		return nil, nil, nil, db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, fmt.Sprintf("unsupported target type %s", target.GetType()))
 	}
 	if err := vet(ctx, target, fieldMaskPaths); err != nil {
 		return nil, nil, nil, db.NoRowsAffected, err
 	}
-
-	if target.GetPublicId() == "" {
-		return nil, nil, nil, db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "missing target public id")
+	projectId := target.GetProjectId()
+	if projectId == "" {
+		t := allocTargetView()
+		t.PublicId = target.GetPublicId()
+		if err := r.reader.LookupByPublicId(ctx, &t); err != nil {
+			return nil, nil, nil, db.NoRowsAffected, errors.Wrap(ctx, err, op, errors.WithMsg(fmt.Sprintf("lookup failed for %s", t.PublicId)))
+		}
+		projectId = t.ProjectId
```

Comment:
We should keep the API consistent with the other methods. We only need this projectId to grab the oplog, but if you check for example `CreateTarget`, we error if there's no projectId straight away (like we did here before this change), and then we just grab the project id from the target and pass it onto the kms wrapper func.

```go
	oplogWrapper, err := r.kms.GetWrapper(ctx, target.GetProjectId(), kms.KeyPurposeOplog)
	if err != nil {
		return nil, nil, nil, errors.Wrap(ctx, err, op, errors.WithMsg("unable to get oplog wrapper"))
	}
```

Reply:
Yea, this makes sense to me because the scopeId is a required field so it shouldn't be missing. I will update the code based on your suggestion.

Reply:
One issue, there is a unit test that actually expects no error message back when the projectId is missing and that's because we fetch the projectId value. The unit test is called `TestRepository_UpdateTcpTarget/empty-project-id-with-name-mask`. I will dig into this and figure out if this unit test case is outdated and can be removed.

Reply:
I went ahead and deleted the unit test because I think it was added by mistake and overlooked in the original PR.

Reply:
Sounds good!